File: pkg/templates/typescript/openai-computer-use/lib/replay.ts
import type { Kernel } from '@onkernel/sdk';

const DEFAULT_REPLAY_GRACE_MS = 5000;
const REPLAY_PROCESSING_DELAY_MS = 2000;
const REPLAY_POLL_TIMEOUT_MS = 60000;
const REPLAY_POLL_INTERVAL_MS = 1000;

export interface ReplayState {
  enabled: boolean;
  replayId: string | null;
  replayViewUrl: string | null;
}

export async function maybeStartReplay(
  kernel: Kernel,
  sessionId: string,
  opts?: { enabled?: boolean },
): Promise<ReplayState> {
  const enabled = opts?.enabled ?? false;
  const state: ReplayState = {
    enabled,
    replayId: null,
    replayViewUrl: null,
  };

  if (!enabled) return state;

  try {
    const replay = await kernel.browsers.replays.start(sessionId);
    state.replayId = replay.replay_id ?? null;
  } catch (error) {
    console.warn(`Warning: failed to start replay recording: ${String(error)}`);
    console.warn('Continuing without replay recording.');
    state.enabled = false;
  }

  return state;
}

export async function maybeStopReplay(
  kernel: Kernel,
  sessionId: string,
  replay: ReplayState,
  opts?: { gracePeriodMs?: number },
): Promise<string | null> {
  if (!replay.enabled || !replay.replayId) return replay.replayViewUrl;

  const gracePeriodMs = opts?.gracePeriodMs ?? DEFAULT_REPLAY_GRACE_MS;
  if (gracePeriodMs > 0) {
    await sleep(gracePeriodMs);
  }

  try {
    await kernel.browsers.replays.stop(replay.replayId, { id: sessionId });
    await sleep(REPLAY_PROCESSING_DELAY_MS);

    const pollStartedAt = Date.now();
    while (Date.now() - pollStartedAt < REPLAY_POLL_TIMEOUT_MS) {
      try {
        const replays = await kernel.browsers.replays.list(sessionId);
        const matchingReplay = replays.find((item) => item.replay_id === replay.replayId);
        if (matchingReplay) {
          replay.replayViewUrl = matchingReplay.replay_view_url ?? null;
          break;
        }
      } catch {
        // Ignore transient polling errors while the replay finishes processing.
      }
      await sleep(REPLAY_POLL_INTERVAL_MS);
    }

    if (!replay.replayViewUrl) {
      console.warn('Warning: replay may still be processing.');
    }
  } catch (error) {
    console.warn(`Warning: failed to stop replay recording cleanly: ${String(error)}`);
  }

  return replay.replayViewUrl;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: pkg/templates/python/openai-computer-use/replay.py
import datetime
import time
from dataclasses import dataclass
from typing import Callable

from kernel import Kernel

DEFAULT_REPLAY_GRACE_SECONDS = 5.0
REPLAY_PROCESSING_DELAY_SECONDS = 2.0
REPLAY_POLL_TIMEOUT_SECONDS = 60.0
REPLAY_POLL_INTERVAL_SECONDS = 1.0


@dataclass
class ReplayState:
    enabled: bool
    replay_id: str | None = None
    replay_view_url: str | None = None


def maybe_start_replay(
    client: Kernel,
    session_id: str,
    enabled: bool = False,
    on_event: Callable[[dict], None] | None = None,
) -> ReplayState:
    state = ReplayState(enabled=enabled)
    if not enabled:
        return state

    started_at = datetime.datetime.now()
    if on_event:
        on_event({"event": "backend", "data": {"op": "browsers.replays.start"}})

    try:
        replay = client.browsers.replays.start(session_id)
        state.replay_id = replay.replay_id
        if on_event:
            on_event(
                {
                    "event": "backend",
                    "data": {
                        "op": "browsers.replays.start.done",
                        "detail": state.replay_id or "",
                        "elapsed_ms": int(
                            (datetime.datetime.now() - started_at).total_seconds() * 1000
                        ),
                    },
                }
            )
    except Exception as exc:
        print(f"Warning: failed to start replay recording: {exc}")
        print("Continuing without replay recording.")
        state.enabled = False

    return state


def maybe_stop_replay(
    client: Kernel,
    session_id: str,
    replay: ReplayState,
    on_event: Callable[[dict], None] | None = None,
    grace_period_seconds: float = DEFAULT_REPLAY_GRACE_SECONDS,
) -> str | None:
    if not replay.enabled or not replay.replay_id:
        return replay.replay_view_url

    if grace_period_seconds > 0:
        time.sleep(grace_period_seconds)

    started_at = datetime.datetime.now()
    if on_event:
        on_event({"event": "backend", "data": {"op": "browsers.replays.stop"}})

    try:
        client.browsers.replays.stop(replay_id=replay.replay_id, id=session_id)
        time.sleep(REPLAY_PROCESSING_DELAY_SECONDS)

        deadline = time.time() + REPLAY_POLL_TIMEOUT_SECONDS
        while time.time() < deadline:
            try:
                replays = client.browsers.replays.list(session_id)
                for replay_item in replays:
                    if replay_item.replay_id == replay.replay_id:
                        replay.replay_view_url = replay_item.replay_view_url
                        break
                if replay.replay_view_url:
                    break
            except Exception:
                pass

            time.sleep(REPLAY_POLL_INTERVAL_SECONDS)

        if on_event:
            on_event(
                {
                    "event": "backend",
                    "data": {
                        "op": "browsers.replays.stop.done",
                        "detail": replay.replay_view_url or replay.replay_id or "",
                        "elapsed_ms": int(
                            (datetime.datetime.now() - started_at).total_seconds() * 1000
                        ),
                    },
                }
            )

        if not replay.replay_view_url:
            print("Warning: replay may still be processing")
    except Exception as exc:
        print(f"Warning: failed to stop replay recording cleanly: {exc}")

    return replay.replay_view_url
